br/pkg/utils/storewatch/watching.go

```diff
@@ -0,0 +1,144 @@
+// the user to listen the events of lifetime of stores.
+package storewatch
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	pd "github.com/tikv/pd/client"
+)
+
+// StoreMeta is the required interface for a watcher.
+// It is striped from pd.Client.
+type StoreMeta interface {
+	// GetAllStores gets all stores from pd.
+	// The store may expire later. Caller is responsible for caching and taking care
+	// of store change.
+	GetAllStores(ctx context.Context, opts ...pd.GetStoreOption) ([]*metapb.Store, error)
+}
+
+// Callback will be called the supported event triggered.
+type Callback interface {
+	OnNewStoreRegistered(store *metapb.Store)
+	OnDisconnect(store *metapb.Store)
+	OnReboot(store *metapb.Store)
+}
+
+// DynCallback is a function based callback set.
+type DynCallback struct {
+	onNewStoreRegistered func(*metapb.Store)
+	onDisconnect         func(*metapb.Store)
+	onReboot             func(*metapb.Store)
+}
+
+// OnNewStoreRegistered will be called once new region added to be watched.
+func (cb *DynCallback) OnNewStoreRegistered(store *metapb.Store) {
+	if cb.onNewStoreRegistered != nil {
+		cb.onNewStoreRegistered(store)
+	}
+}
+
+// OnDisconnect will be called once the store is disconnected.
+func (cb *DynCallback) OnDisconnect(store *metapb.Store) {
+	if cb.onDisconnect != nil {
+		cb.onDisconnect(store)
+	}
+}
+
+// OnReboot will be called once the store is rebooted.
+func (cb *DynCallback) OnReboot(store *metapb.Store) {
+	if cb.onReboot != nil {
+		cb.onReboot(store)
+	}
+}
+
+// DynCallbackOpt is the option for DynCallback.
+type DynCallbackOpt func(*DynCallback)
+
+// WithOnNewStoreRegistered adds a hook to the callback.
+func WithOnNewStoreRegistered(f func(*metapb.Store)) DynCallbackOpt {
+	return func(cb *DynCallback) {
+		cb.onNewStoreRegistered = f
+	}
+}
+
+// WithOnDisconnect adds a hook to the callback.
+func WithOnDisconnect(f func(*metapb.Store)) DynCallbackOpt {
+	return func(cb *DynCallback) {
+		cb.onDisconnect = f
+	}
+}
+
+// WithOnReboot adds a hook to the callback.
+func WithOnReboot(f func(*metapb.Store)) DynCallbackOpt {
+	return func(cb *DynCallback) {
+		cb.onReboot = f
+	}
+}
+
+// MakeCallback creates a callback with the given options.
+// Allowed options: WithOnNewStoreRegistered, WithOnDisconnect, WithOnReboot.
+func MakeCallback(opts ...DynCallbackOpt) Callback {
+	cb := &DynCallback{}
+	for _, opt := range opts {
+		opt(cb)
+	}
+	return cb
+}
+
+// Watcher watches the lifetime of stores.
+// generally it should be advanced by calling the `Step` call.
+type Watcher struct {
+	cli StoreMeta
+	cb  Callback
+
+	lastStores map[uint64]*metapb.Store
+}
+
+func New(cli StoreMeta, cb Callback) *Watcher {
+	return &Watcher{
+		cli:        cli,
+		cb:         cb,
+		lastStores: make(map[uint64]*metapb.Store),
+	}
+}
+
+func (w *Watcher) Step(ctx context.Context) error {
+	liveStores, err := w.cli.GetAllStores(ctx, pd.WithExcludeTombstone())
```

Comment:
It's better to use `GetAllTiKVStoresWithRetry` to skip tiflash node and add the retry mechanism